controllers/progressivesync_controller.go

```diff
@@ -341,168 +281,265 @@ func (r *ProgressiveSyncReconciler) getOwnedAppsFromClusters(ctx context.Context
-func (r *ProgressiveSyncReconciler) syncApp(ctx context.Context, appName string) (*argov1alpha1.Application, error) {
+func (r *ProgressiveSyncReconciler) syncApp(ctx context.Context, app argov1alpha1.Application) error {
 	syncReq := applicationpkg.ApplicationSyncRequest{
-		Name: &appName,
+		Name: &app.Name,
 	}
 
-	return r.ArgoCDAppClient.Sync(ctx, &syncReq)
-}
-
-// updateStatusWithRetry updates the progressive sync object status with backoff
-func (r *ProgressiveSyncReconciler) updateStatusWithRetry(ctx context.Context, pr *syncv1alpha1.ProgressiveSync) error {
-
-	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-
-		key := client.ObjectKeyFromObject(pr)
-		latest := syncv1alpha1.ProgressiveSync{}
-		if err := r.Client.Get(ctx, key, &latest); err != nil {
-			return err
-		}
-		latest.Status = pr.Status
-		if err := r.Client.Status().Update(ctx, &latest); err != nil {
-			return err
-		}
-		return nil
+	_, err := r.ArgoCDAppClient.Sync(ctx, &syncReq)
+	if err != nil && !strings.Contains(err.Error(), "another operation is already in progress") {
+		return err
+	}
 
-	})
-	return retryErr
+	return nil
 }
 
-// markAppAsSynced marks the app as synced in the specified stage
-func (r *ProgressiveSyncReconciler) markAppAsSynced(ctx context.Context, app argov1alpha1.Application, stage syncv1alpha1.ProgressiveSyncStage, pss utils.ProgressiveSyncState) error {
-
-	log := r.Log.WithValues("stage", stage.Name, "app", fmt.Sprintf("%s/%s", app.Namespace, app.Name))
-
-	pss.MarkAppAsSynced(app, stage)
+// reconcile performs the actual reconciliation logic
+func (r *ProgressiveSyncReconciler) reconcile(ctx context.Context, ps syncv1alpha1.ProgressiveSync) (syncv1alpha1.ProgressiveSync, ctrl.Result, error) {
 
-	log.Info("app marked as synced")
+	log := log.FromContext(ctx)
 
-	return nil
-}
+	// Observe ProgressiveSync generation
+	if ps.Status.ObservedGeneration != ps.Generation {
+		ps.Status.ObservedGeneration = ps.Generation
+		return syncv1alpha1.ProgressiveSyncProgressing(ps), ctrl.Result{Requeue: true}, nil
+	}
 
-// reconcileStage reconcile a ProgressiveSyncStage
-func (r *ProgressiveSyncReconciler) reconcileStage(ctx context.Context, ps syncv1alpha1.ProgressiveSync, stage syncv1alpha1.ProgressiveSyncStage, pss utils.ProgressiveSyncState) (syncv1alpha1.ProgressiveSync, reconcile.Result, error) {
-	log := r.Log.WithValues("progressivesync", fmt.Sprintf("%s/%s", ps.Namespace, ps.Name), "applicationset", ps.Spec.SourceRef.Name, "stage", stage.Name, "syncedAtStage", pss)
+	// Create the progressive sync state configmap if it doesn't exist
+	if err := r.CreateStateMap(ctx, ps); err != nil {
+		log.Error(err, "unable to create state map")
+		return ps, ctrl.Result{RequeueAfter: RequeueDelayOnError}, err
+	}
 
-	// Get the clusters to update
-	clusters, err := r.getClustersFromSelector(ctx, stage.Targets.Clusters.Selector)
+	// Read the latest state
+	state, err := r.ReadStateMap(ctx, ps)
```

Comment:
why not read directly and create the statemap if it doesn't exist directly?

Reply:
I think because this is how it came along :)

At the moment we need to always call create before reading, otherwise we might want to try to read something that doesn't exist. I can change the logic here

Reply:
Done